modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/TableScan.java

```diff
@@ -0,0 +1,245 @@
+    }
+
+    /**
+     * Table scan iterator.
+     */
+    public static class IteratorImpl extends GridCloseableIteratorAdapter<Object[]> {
+        /** */
+        private final ExecutionContext ectx;
+
+        /** */
+        private final TableDescriptor desc;
+
+        /** */
+        private int cacheId = UNDEFINED_CACHE_ID;
+
+        /** */
+        private GridCacheContext<?,?> cctx;
+
+        /** */
+        private Queue<GridDhtLocalPartition> parts;
+
+        /** */
+        private GridDhtLocalPartition part;
+
+        /** */
+        private GridCursor<? extends CacheDataRow> cur;
+
+        /** */
+        private Object[] next;
+
+        /** */
+        public IteratorImpl(ExecutionContext ectx, TableDescriptor desc) {
+            this.ectx = ectx;
+            this.desc = desc;
+        }
+
+        /** {@inheritDoc} */
+        @Override protected Object[] onNext() {
+            if (next == null)
+                throw new NoSuchElementException();
+
+            Object[] next = this.next;
+
+            this.next = null;
+
+            return next;
+        }
+
+        /** {@inheritDoc} */
+        @Override protected boolean onHasNext() throws IgniteCheckedException {
+            assert cctx != null && parts != null;
+
+            if (next != null)
+                return true;
+
+            while (true) {
+                if (cur == null) {
+                    if ((part = parts.poll()) == null)
+                        break;
+
+                    IgniteCacheOffheapManager.CacheDataStore ds = part.dataStore();
+
+                    // TODO introduce MvccSnapshot to the table iterator.
+                    cur = cacheId == UNDEFINED_CACHE_ID ? ds.cursor() : ds.cursor(cacheId, null);
+                }
+
+                if (cur.next()) {
+                    CacheDataRow row = cur.get();
+
+                    if (!desc.matchType(row))
+                        continue;
+
+                    next = desc.toRow(ectx, cctx, row);
+
+                    break;
+                } else {
+                    cur = null;
+
+                    part.release();
+                    part = null;
+                }
+            }
+
+            return next != null;
+        }
+
+        /** {@inheritDoc} */
+        @Override protected void onClose() {
+            if (part != null)
+                part.release();
+
+            part = null;
+
+            while (!F.isEmpty(parts))
+                parts.poll().release();
+
+            parts = null;
+        }
+
+        /** */
+        public Iterator<Object[]> init() {
+            if (isClosed())
+                return Collections.emptyIterator();
+
+            cctx = ectx.parent().kernal().cache().context().cacheContext(cacheId = desc.cacheId());
+
+            if (!cctx.group().sharedGroup())
+                cacheId = UNDEFINED_CACHE_ID;
+
+            GridDhtPartitionTopology top = cctx.topology();
+            top.readLock();
```

Comment:
Anyway locking is needed to check whether the interested cache is still there, additionally it saves us from exceptions in case a topology is changed while we reserve partitions.